server/resp/src/main/java/org/infinispan/server/resp/scripting/LuaTaskEngine.java

```diff
@@ -0,0 +1,380 @@
+ */
+public class LuaTaskEngine implements TaskEngine {
+   private final LuaContextPool pool;
+   private final ScriptingManager scriptingManager;
+
+   public LuaTaskEngine(ScriptingManager scriptingManager) {
+      this.scriptingManager = scriptingManager;
+      this.pool = new LuaContextPool(LuaContext::new, 2, 4);
+   }
+
+   public void shutdown() {
+      pool.shutdown();
+   }
+
+   public CompletionStage<Void> eval(Resp3Handler handler, ChannelHandlerContext ctx, String code, String[] keys, String[] args, long flags) {
+      return handler.getBlockingManager().supplyBlocking(() -> {
+         LuaContext luaCtx = pool.borrow();
+         LuaCode script = scriptLoad(code, false);
+         try {
+            luaCtx.registerScript(script);
+            runScript(luaCtx, handler, ctx, script, keys, args, flags);
+         } finally {
+            luaCtx.unregisterScript(script);
+         }
+         return luaCtx;
+      }, "eval").thenApplyAsync(luaCtx -> {
+         // Process the lua object on the stack and send it to the actual writer
+         luaToResp(luaCtx.lua, handler);
+         // Pop the error handler
+         luaCtx.lua.pop(1);
+         // Return the lua context to the pool
+         pool.returnToPool(luaCtx);
+         return null;
+      }, ctx.channel().eventLoop());
+   }
+
+   public CompletionStage<Void> evalSha(Resp3Handler handler, ChannelHandlerContext ctx, String sha, String[] keys, String[] args, long flags) {
+      return handler.getBlockingManager().supplyBlocking(() -> {
+         LuaContext luaCtx = pool.borrow();
+         ScriptWithMetadata script;
+         try {
+            script = scriptingManager.getScriptWithMetadata(scriptName(sha.toUpperCase()));
+         } catch (CacheException e) {
+            throw new RuntimeException("NOSCRIPT No matching script. Please use EVAL.");
+         }
+         LuaCode code = LuaCode.fromScript(script);
+         luaCtx.registerScript(code);
+         runScript(luaCtx, handler, ctx, code, keys, args, flags);
+         return luaCtx;
+      }, "evalsha").thenApplyAsync(luaCtx -> {
+         // Process the lua object on the stack and send it to the actual writer
+         luaToResp(luaCtx.lua, handler);
+         // Pop the error handler
+         luaCtx.lua.pop(1);
+         pool.returnToPool(luaCtx);
+         return null;
+      }, ctx.channel().eventLoop());
+   }
+
+   private void runScript(LuaContext luaCtx, Resp3Handler handler, ChannelHandlerContext ctx, LuaCode script, String[] keys, String[] args, long flags) {
+      luaCtx.handler = handler;
+      luaCtx.ctx = ctx;
+      luaCtx.flags = flags;
+      luaCtx.flags |= script.flags();
+      ResponseWriter oldWriter = handler.writer(new LuaResponseWriter(luaCtx.lua));
+      try {
+         TransactionContext.startTransactionContext(ctx);
+         runScript(luaCtx.lua, script, keys, args);
+      } catch (Throwable t) {
+         throw new RuntimeException(t);
```

Comment:
This RuntimeException seems extraneous. I also worry that if an exception does happen, nothing is propagated to the user as the handler isn't written to.

Reply:
With the change in the callers, this should be handled properly now